### PR TITLE
Fast USDC CLI: Transfer

### PR DESCRIPTION
**Pull request by samsiegart**
refs https://github.com/Agoric/agoric-sdk/issues/10339

## Description

- Adds a `config` command for setting all the required variables relating to IBC channels, mnemonics, rpc endpoints, etc.
- Adds a `transfer` command for simulating the Noble Express frontend. It accepts an end user cosmos address destination, and a USDC amount as args. TODO: It should query the agoric Fast USDC LCA address, but for now it just stubs a hardcoded address there, because I'm not sure what the query should look like.
- Also TODO, add a command to query the status of pending transfers

### Documentation Considerations
Tried to make the `help` command as self-documenting as possible

### Testing Considerations
Manually tested on testnets with testnet tokens that:
- It queries the noble forwarding address for the "agoricXXX?EUD=dydxYYY" address 
- It registers the forwarding address on noble if it doesn't exist yet (I tested on noble mainnet for this part actually, not testnet)
- It sends a depositForBurn txn on ethereum (I used sepolia testnet) with the correctly encoded noble address: https://sepolia.etherscan.io/tx/0xde6d7ea6a6d2737e67524dca70372610c5ac0476d75d4c30fc4270e89e36de77

I put the config used for that test run in `demo/testnet` for reference.

Added unit tests as well.

### Upgrade Considerations
The CLI is a client program and doesn't need to be upgraded in production.

### CODE REVIEW

**Comment by dckc** [comment]
> didn't unit test "transfer" yet.

Is there some particular reason to land this without such a test?

**Comment by samsiegart** [comment]
>Is there some particular reason to land this without such a test?

Are you blocked on https://github.com/Agoric/agoric-sdk/pull/10441? It would be better to merge this sooner than later so that we could make that PR fit the structure of this one.

If it's not a blocker, I can add the tests if this looks good to you so far. I mostly wanted to make sure my approach was agreeable before doing all that extra work. For example, removing mock-fs was significant feedback, and I might've been inclined to use a monkey-patching approach on the rest of the stuff otherwise. I've gone ahead and removed mock-fs though, and used file objects instead as discussed, so PTAL @dckc

File: packages/fast-usdc/src/util/cctp.js
Comment (dckc) [inline packages/fast-usdc/src/util/cctp.js]: How about passing in a provider instead of a string?

Just like passing a file object instead of a filename.

File: packages/fast-usdc/src/util/agoric.js
Comment (dckc) [inline packages/fast-usdc/src/util/agoric.js]: Let's pass a vstorage client rather than a string.

**Review by dckc — COMMENTED** [review]
2 more suggestions similar to file objects instead of filenames.

File: packages/fast-usdc/src/util/cctp.js
Comment (dckc) [reply packages/fast-usdc/src/util/cctp.js]: This should make unit testing transfer straightforward

**Comment by dckc** [comment]
> > Is there some particular reason to land this without such a test?
> 
> Are you blocked on #10441?

No; #10441 is just for discussion. It's a bunch of suggestions about an approach for testing and such.

> If it's not a blocker, I can add the tests if this looks good to you so far.

Sure, a lot of it looks good.

File: packages/fast-usdc/src/util/noble.js
Comment (dckc) [inline packages/fast-usdc/src/util/noble.js]: **note** (something I hope to do after this PR):

While this does support OCap discipline by letting the caller override the `get` parameter, `get` has excess authority: it's access to the whole web. [makeLCD](https://github.com/Agoric/agoric-sdk/blob/9ffdc26a7466d0ac9ad786dae07922e0d1984ec8/packages/fast-usdc/src/cli/cosmos-api.js#L16) combines `fetch` and an API url to make something with all and only the authority to use the Cosmos REST API at one endpoint. Here's hoping we manage to factor that out in due course.

File: packages/fast-usdc/src/util/file.js
Comment (dckc) [inline packages/fast-usdc/src/util/file.js:14]: **suggestion**: With more than 3 args, naming them tends to help. It works nicely with `Pick<typeof fs, 'readFile' | 'writeFile' | ...>` too.

File: packages/fast-usdc/src/util/noble.js
Comment (dckc) [inline packages/fast-usdc/src/util/noble.js]: `nobleRpc` should be an RPC client object rather than a string. Something like this should work, based on [some endo playground experience](https://github.com/endojs/playground/blob/main/packages/ag-trade/src/smartWallet.js#L200-L217):

Use [makeTendermintRpcClient(url, fetch)](https://github.com/Agoric/agoric-sdk/blob/d19b293d5b28a2d78754eb7912771bdb66d5fb37/packages/casting/src/makeHttpClient.js#L28C14-L28C37) from casting and the rest comes from cosmjs:

```js
import { Tendermint34Client } from '@cosmjs/tendermint-rpc';

...

  const cometClient = await Tendermint34Client.create(rpcClient);
  const clientWithSigner = await SigningStargateClient.createWithSigner(
    cometClient,
    ...
```

File: packages/fast-usdc/src/cli/cli.js
Comment (dckc) [inline packages/fast-usdc/src/cli/cli.js:52]: most of the code doesn't seem to need write access.

Consider grabbing [makeFileRW and makeFileRd](https://github.com/Agoric/agoric-3-proposals/blob/4f553e0e084ec7a8203b7b5735d9ec0dfe2a57a7/packages/synthetic-chain/src/lib/webAsset.js) from synthetic-chain. make a RW and then call `.readOnly()`; only pass the RW to the things that need it.

Comment (dckc) [reply packages/fast-usdc/src/cli/cli.js:52]: we can do that later. not critical.

File: packages/fast-usdc/src/util/cctp.js
Comment (samsiegart) [reply packages/fast-usdc/src/util/cctp.js]: Done

File: packages/fast-usdc/src/util/agoric.js
Comment (samsiegart) [reply packages/fast-usdc/src/util/agoric.js]: Done

File: packages/fast-usdc/src/util/noble.js
Comment (samsiegart) [reply packages/fast-usdc/src/util/noble.js]: Ack

File: packages/fast-usdc/src/util/noble.js
Comment (samsiegart) [reply packages/fast-usdc/src/util/noble.js]: Why though? It adds more dependencies and lines of code, while the `SigningStargateClient` interface supports a string

File: packages/fast-usdc/src/cli/cli.js
Comment (samsiegart) [reply packages/fast-usdc/src/cli/cli.js:52]: Ack

**Comment by samsiegart** [comment]
Added test coverage for the full transfer flow now

File: packages/fast-usdc/src/util/noble.js
Comment (dckc) [reply packages/fast-usdc/src/util/noble.js]: > It adds more dependencies

It adds lines in `package.json` maybe, but the dependencies are already there, indirectly.

> `SigningStargateClient` interface supports a string

yes, but that string separates designation from authority. If we pass a string, `SigningStargateClient` is going to use ambient access to the network - `fetch` or the like. How would we test `registerFwdAccount`? Perhaps there are tools that use global mutable state to override `fetch`. That sort of spooky action-at-a-distance is difficult to manage at scale. And it means `registerFwdAccount` doesn't support patterns of reuse such as passing in an RPCClient with built-in rate-limiting.

File: packages/fast-usdc/test/cli/transfer.test.ts
Comment (dckc) [inline packages/fast-usdc/test/cli/transfer.test.ts]: This is a pretty important external interface. Consider using `t.snapshot()` rather than `t.deepEqual()`; the snapshot markdown files are particularly useful docs. Plus, they're low maintenance (just run `yarn test -u` and check the diffs).

File: packages/fast-usdc/test/cli/transfer.test.ts
Comment (dckc) [inline packages/fast-usdc/test/cli/transfer.test.ts]: That allows the caller to modify `txnArgs`.

```suggestion
  return { provider, getTxnArgs: () => harden([...txnArgs]) };
```

provider should be hardened before it's returned too. Perhaps not critical in test code, but it's only a few characters and it's a good habit.

File: packages/fast-usdc/testing/mocks.ts
Comment (dckc) [inline packages/fast-usdc/testing/mocks.ts]: `getSigned` lets callers modify `signedArgs`.

**suggest**: Harden at the earliest opportunity:

```suggestion
      signedArgs = harden(args);
```

File: packages/fast-usdc/src/cli/transfer.js
Comment (dckc) [inline packages/fast-usdc/src/cli/transfer.js]: The `+` syntax isn't defined by this module, is it? It seems like we should be able to import a function that combines these two.

I **suggest** making a local function and marking it `TODO: update syntax to sync with other components` or something.

Are we even still using `+` syntax? In something linked from **Fast USDC Eng Design / Sketch**, I see:

> ❓ [How should the parameters be separated from the address?](https://codecider-app.vercel.app/d/d8ed6dff-6afd-43e5-8a60-4894d1e693a4#?-how-should-the-parameters-be-separated-from-the-address)
> ✓ Use "?" separator
> ✓ [How should the parameters be encoded?](https://codecider-app.vercel.app/d/d8ed6dff-6afd-43e5-8a60-4894d1e693a4#?!-how-should-the-parameters-be-encoded)
> ✓ Use URL query param encoding

File: packages/fast-usdc/src/cli/transfer.js
Comment (dckc) [inline packages/fast-usdc/src/cli/transfer.js:44]: **note**: `fetch` for all the web is a more authority than is needed; an attenuated fetch that can only get JSON on URLs under `nobleApi` seems necessary and sufficient. IOU an issue/PR to add that to `@agoric/client-utils` and use it here.

File: packages/fast-usdc/src/util/noble.js
Comment (dckc) [inline packages/fast-usdc/src/util/noble.js]: **request**: add a docstring to cite docs for the x/forwarding module. a `@file` thing or on one or more of these functions.

something like:

https://github.com/noble-assets/forwarding/blob/9d7657a/proto/noble/forwarding/v1/query.proto
v2.0.0 10 Nov 2024

File: packages/fast-usdc/src/util/noble.js
Comment (dckc) [inline packages/fast-usdc/src/util/noble.js:105]:
```suggestion
  /** @type {{ address: string, exists: boolean }} */
  const { address, exists } = forwardingAddressRes;
```

It might be bettter on a `@returns` for the function.

File: packages/fast-usdc/src/cli/transfer.js
Comment (dckc) [inline packages/fast-usdc/src/cli/transfer.js]: `return` suggests failing to register is a success condition. Help me understand that?

It looks like a **silent failure**. It's logged, so it's not entirely silent. But the caller doesn't get the usual signal about success / failure. The unix exit code should be non-zero, right?

File: packages/fast-usdc/src/cli/transfer.js
Comment (dckc) [inline packages/fast-usdc/src/cli/transfer.js]: likewise - a broken promise from `depositForBurn` is success?
or another **silent failure**?

File: packages/fast-usdc/src/util/cctp.js
Comment (dckc) [inline packages/fast-usdc/src/util/cctp.js:6]: :clap: 

Maybe one README somewhere should cite noble docs instead of putting them on individual functions / files?
I don't feel strongly; but I suggest: whenever you look something up and you find an authoritative answer, stick the URL (and maybe title, date, git hash, ...) in a comment.

File: packages/fast-usdc/src/util/cctp.js
Comment (dckc) [inline packages/fast-usdc/src/util/cctp.js:25]: Where did these ABIs come from?

ok... I see `contractAbi` on https://developers.circle.com/interactive-quickstarts/cctp . Seems to match.

I presume `tokenAbi` is from ERC20.

File: packages/fast-usdc/src/cli/transfer.js
Comment (dckc) [inline packages/fast-usdc/src/cli/transfer.js:66]: I think the indirection here only obscures things. **suggestion**:

```suggestion
    ethProvider ||= new ethers.JsonRpcProvider(config.ethRpc);
```

File: packages/fast-usdc/src/util/cctp.js
Comment (dckc) [inline packages/fast-usdc/src/util/cctp.js:28]: As noted in `transfer.js`, I think inlining this is better. **suggest**: please let's punt this.

```suggestion
```

`makeProvider` breaks our "modules should not export powerful objects (for example, objects that close over fs or process.env) " [guideline](https://github.com/Agoric/agoric-sdk/wiki/OCap-Discipline)

File: packages/fast-usdc/src/util/cctp.js
Comment (dckc) [inline packages/fast-usdc/src/util/cctp.js]: **suggestion**: hoist to a manifest constant:

```js
const USDC_DECIMALS= 6;
```

File: packages/fast-usdc/test/cli/transfer.test.ts
Comment (dckc) [inline packages/fast-usdc/test/cli/transfer.test.ts:151]: I wonder how to review that. I suppose we'll eventually do integration testing, and we'll find out real quick if it's not right at that point.

Meanwhile, I did try to puzzle it out a little, and this form might be slightly more legible. Dunno whether it's really useful, but I'll share it rather than throw it away:

```suggestion
  t.deepEqual(ethers.decodeRlp(mockEthProvider.getTxnArgs()[0]), [
    '0x',
    '0x01',
    '0x',
    '0x1c7d4b196cb0c7b01d743fbc6116a902379c7238',
    '0x',
    '0x095ea7b30000000000000000000000009f3b8679c73c2fef8b59b4f3444d4e156fb70aa50000000000000000000000000000000000000000000000000000000008f0d180',
    '0x011a',
    '0xb2d87eeb1cb36243f95662739e2a7bd4bddc2b8afe189ac4848ec71cc314335b',
    '0x68136695c644f69474e2e30ea7059f9b380fbb1a09beb3580f73d3ea349912ab',
  ]);
```

File: packages/fast-usdc/test/util/file.test.ts
Comment (dckc) [inline packages/fast-usdc/test/util/file.test.ts:63]: this file of tests for `makeFile` seems like overkill. Sorry if I suggested otherwise. I suppose they don't hurt much, other than maintenance.

File: packages/fast-usdc/src/cli/cli.js
Comment (dckc) [inline packages/fast-usdc/src/cli/cli.js:25]: is it useful to inject these 2 any more? Don't we just inject `fetch` and `makeSigner` and such?

**note**: ok... I get it... we mock `transferHelpers` to check that `initProgram` calls `transferHelpers` as expected and then unit test `transferHelpers` such as `.transfer`. fair enough.

File: packages/fast-usdc/src/cli/transfer.js
Comment (dckc) [inline packages/fast-usdc/src/cli/transfer.js]: **suggestion**: use `@import` rather than `@typedef`. `@typedef` loses docstrings.

File: packages/fast-usdc/src/cli/transfer.js
Comment (dckc) [inline packages/fast-usdc/src/cli/transfer.js:20]: The rest of agoric-sdk uses `@param`. This is more concise in a way, so I don't mind too much. It's also more like typescript, which we seem to be headed toward. So maybe it's better.

**Review by dckc — CHANGES_REQUESTED** [review]
It mostly looks great, but...

It looks like it gives a successful exit code (0) on failure, which makes it unsuitable as an integration testing tool. See **silent failure** comments.

Consider the **suggestion**s entirely at your discretion; feel free to ignore entirely, decline, adopt, adapt, etc.

File: packages/fast-usdc/src/cli/transfer.js
Comment (samsiegart) [reply packages/fast-usdc/src/cli/transfer.js]: Ooh TIL... done.

File: packages/fast-usdc/src/cli/transfer.js
Comment (samsiegart) [reply packages/fast-usdc/src/cli/transfer.js:20]: Ack, if you don't feel strongly I'll leave as is

File: packages/fast-usdc/src/cli/transfer.js
Comment (samsiegart) [reply packages/fast-usdc/src/cli/transfer.js]: Fixed, after discussion we're going with `?EUD=`, good catch.

File: packages/fast-usdc/src/cli/transfer.js
Comment (samsiegart) [reply packages/fast-usdc/src/cli/transfer.js:44]: Ack

File: packages/fast-usdc/src/cli/transfer.js
Comment (samsiegart) [reply packages/fast-usdc/src/cli/transfer.js]: No you're right, it shouldn't be considered a success condition. I just went back and made sure all these fatal errors actually throw.

File: packages/fast-usdc/src/cli/transfer.js
Comment (samsiegart) [reply packages/fast-usdc/src/cli/transfer.js:66]: I just didn't want transfer.js to have to depend on ethers. It would be nice if all that domain specific stuff was handled by the util

File: packages/fast-usdc/src/cli/transfer.js
Comment (samsiegart) [reply packages/fast-usdc/src/cli/transfer.js]: Fixed here as well

File: packages/fast-usdc/src/util/cctp.js
Comment (samsiegart) [reply packages/fast-usdc/src/util/cctp.js:6]: I'm not sure what else I'd put in the README at this point, I think it's alright to just leave this as a doc comment

File: packages/fast-usdc/src/util/cctp.js
Comment (samsiegart) [reply packages/fast-usdc/src/util/cctp.js:25]: Yea and from looking at https://sepolia.etherscan.io/address/0x9f3b8679c73c2fef8b59b4f3444d4e156fb70aa5#writeContract.

I thought this would be snappier than importing the whole json ABI, which can be found [here](https://github.com/circlefin/evm-cctp-contracts/blob/e4e6e2fccd6820002eb4a5b4fabdc8ea11031ad9/docs/abis/cctp/TokenMessenger.json)

File: packages/fast-usdc/src/util/cctp.js
Comment (samsiegart) [reply packages/fast-usdc/src/util/cctp.js:28]: I guess by powerful you mean because it has network access? That rule is already broken by `ethers` itself though right? Otherwise it would be nice to be able to pass `fetch` etc. explicitly. I think it's just cleaner to have all the `ethers` stuff in one file, it will be easier to swap out or modify if needed.

File: packages/fast-usdc/src/util/cctp.js
Comment (samsiegart) [reply packages/fast-usdc/src/util/cctp.js]: Done

File: packages/fast-usdc/src/util/noble.js
Comment (samsiegart) [reply packages/fast-usdc/src/util/noble.js]: Added a doc comment

File: packages/fast-usdc/src/util/file.js
Comment (samsiegart) [reply packages/fast-usdc/src/util/file.js:14]: The args have gotten long in quite a few places because of the need for dependency injection, so I'll look through this last thing before merging to see if I can clean it up some.

File: packages/fast-usdc/test/cli/transfer.test.ts
Comment (samsiegart) [reply packages/fast-usdc/test/cli/transfer.test.ts]: Done

File: packages/fast-usdc/test/cli/transfer.test.ts
Comment (samsiegart) [reply packages/fast-usdc/test/cli/transfer.test.ts]: Done

File: packages/fast-usdc/test/cli/transfer.test.ts
Comment (samsiegart) [reply packages/fast-usdc/test/cli/transfer.test.ts:151]: Yea I'm not really sure either. Just left a comment because I found a website that can decode it. But I spent a little bit trying to do it myself with `ethers` and couldn't succeed.

File: packages/fast-usdc/test/util/file.test.ts
Comment (samsiegart) [reply packages/fast-usdc/test/util/file.test.ts:63]: Np

File: packages/fast-usdc/testing/mocks.ts
Comment (samsiegart) [reply packages/fast-usdc/testing/mocks.ts]: Done

**Review by samsiegart — COMMENTED** [review]
PTAL

File: packages/fast-usdc/src/util/noble.js
Comment (samsiegart) [reply packages/fast-usdc/src/util/noble.js:105]: Done

File: packages/fast-usdc/src/util/cctp.js
Comment (dckc) [reply packages/fast-usdc/src/util/cctp.js:28]: > I guess by powerful you mean because it has network access?

yes. we only seem to use it as a fallback in cases where the caller can explicitly supply a provider, so I guess it doesn't do much harm. Want to add a comment to warn folks not to use it otherwise? (maybe not worth another round of ci. I won't insist)

File: packages/fast-usdc/test/cli/transfer.test.ts
Comment (dckc) [reply packages/fast-usdc/test/cli/transfer.test.ts:151]: cool. the comment is quite nice.

File: packages/fast-usdc/src/cli/transfer.js
Comment (dckc) [inline packages/fast-usdc/src/cli/transfer.js:35]: This implementation assumes `destination` doesn't need URL encoding. I suppose that's true for any correct bech32 addresses. But I'm not sure we validate that `destination` is a correct bech32 address.

`packages/fast-usdc/src/utils/address.js` has some `addressTools`. I wonder if a function there to combine these is worthwhile. I'm OK postponing until the need is more evident.

File: packages/fast-usdc/src/util/noble.js
Comment (dckc) [inline packages/fast-usdc/src/util/noble.js:93]: I meant a docstring on `queryForwardingAccount`. sorry i wasn't clear. not critical.

**Review by dckc — APPROVED** [review]
:shipit:

**Comment by samsiegart** [comment]
Logs from a successful run after updating the URL encoding (ignore the insufficient funds thing, that's just because my eth account was empty. Also, the "agoric1234asdadfga" address was stubbed before running):
```
$ fast-usdc transfer 0.25 dydx1234 --home=demo/testnet/
forwarding destination agoric1234asdadfga?EUD=dydx1234
querying forward address details from noble api: https://noble-api.polkachu.com/noble/forwarding/v1/address/channel-21/agoric1234asdadfga%3FEUD%3Ddydx1234/
got forwarding address details: {"address":"noble1s7jcagjg34z9ucrfhkcg9t880ttksjsrgs8amq","exists":false}
got noble wallet from seed
registering fwd account on noble
signing message Object <[Object: null prototype] {}> {
  typeUrl: '/noble.forwarding.v1.MsgRegisterAccount',
  value: Object <[Object: null prototype] {}> {
    signer: 'noble1n4mqetruv26lm2frkjah86h642ts84qy9hfyuv',
    recipient: 'agoric1234asdadfga?EUD=dydx1234',
    channel: 'channel-21'
  }
}
Transaction successful with hash: 5E06175CFEE6B053467B458FAE07037375710849F44F75FD3586D07A2B726A92
approving
SES_UNHANDLED_REJECTION: (Error#1)
Error#1: insufficient funds for intrinsic transaction cost (transaction="0x02f8b183aa36a7058310528484139136b482da0b941c7d4b196cb0c7b01d743fbc6116a902379c723880b844095ea7b30000000000000000000000009f3b8679c73c2fef8b59b4f3444d4e156fb70aa5000000000000000000000000000000000000000000000000000000000003d090c080a029509d5664e90049a836cba1a1213910ecc2c117f079b4382f0cc01b310bd609a05a2c351e7876fbf829afb5d5ee27c9f27c9b743048675db84d1b08e553c589bd", info={ "error": { "code": -32000, "message": "INTERNAL_ERROR: insufficient funds" } }, code=INSUFFICIENT_FUNDS, version=6.13.4)
  at makeError (file:///home/samsiegart/agoric-sdk/node_modules/ethers/lib.esm/utils/errors.js:124:21)
  at JsonRpcProvider.getRpcError (file:///home/samsiegart/agoric-sdk/node_modules/ethers/lib.esm/providers/provider-jsonrpc.js:698:24)
  at file:///home/samsiegart/agoric-sdk/node_modules/ethers/lib.esm/providers/provider-jsonrpc.js:298:45
  at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
  ```

File: packages/fast-usdc/src/cli/transfer.js
Comment (samsiegart) [reply packages/fast-usdc/src/cli/transfer.js:35]: I had to fix this anyway. It will now encode all of `appendedAddr`, which I think should work regardless.

File: packages/fast-usdc/src/util/noble.js
Comment (samsiegart) [reply packages/fast-usdc/src/util/noble.js:93]: I would but then it makes me pull all the args into the docstring as well, quite annoying

Comment (dckc) [reply packages/fast-usdc/src/util/noble.js:93]: oh... if using `@type` rather than `@param` means no prose docstrings, that's going to be a problem real soon.

File: packages/fast-usdc/src/cli/transfer.js
Comment (turadg) [reply packages/fast-usdc/src/cli/transfer.js:20]: LGTM. We might have the whole package on `.ts` soon: https://github.com/Agoric/agoric-sdk/pull/10480

**Pull request merged**
